Classes/Private/Manager/Audio/OCTAudioEngine.m

```diff
@@ -83,459 +38,162 @@ - (instancetype)init
-
-    if (status != noErr) {
-        [self fillError:error
-               withCode:status
-            description:@"AUGraphStop"
-          failureReason:@"Failed to stop graph"];
+    if ([self.inputQueue setDeviceID:inputDeviceID error:error]) {
+        _inputDeviceID = inputDeviceID;
+        return YES;
+    }
+    else {
         return NO;
     }
-
-    TPCircularBufferClear(&_outputBuffer);
-    TPCircularBufferClear(&_inputBuffer);
-
-#if TARGET_OS_IPHONE
-    AVAudioSession *session = [AVAudioSession sharedInstance];
-
-    return [session setActive:NO error:error];
-#else
-#warning TODO audio OSX
-    return NO;
 #endif
 }
 
-- (BOOL)routeAudioToSpeaker:(BOOL)speaker error:(NSError **)error;
+- (BOOL)setOutputDeviceID:(NSString *)outputDeviceID error:(NSError **)error
 {
 #if TARGET_OS_IPHONE
+    _outputDeviceID = outputDeviceID;
     AVAudioSession *session = [AVAudioSession sharedInstance];
 
-    AVAudioSessionPortOverride override = (speaker) ? AVAudioSessionPortOverrideSpeaker : AVAudioSessionPortOverrideNone;
+    AVAudioSessionPortOverride override;
+    if (outputDeviceID == OCTOutputDeviceSpeaker) {
+        override = AVAudioSessionPortOverrideSpeaker;
+    }
+    else {
+        override = AVAudioSessionPortOverrideNone;
+    }
 
     return [session overrideOutputAudioPort:override error:error];
 #else
-#warning TODO audio OSX
-    return NO;
-#endif
-}
-
-#pragma mark - Audio Status
-- (BOOL)isAudioRunning:(NSError **)error
-{
-    Boolean running = false;
-    OSStatus status = _AUGraphIsRunning(self.processingGraph,
-                                        &running);
-    if (status != noErr) {
-        [self fillError:error
-               withCode:status
-            description:@"Check if Audio Graph is running"
-          failureReason:@"Failed to check if graph is running"];
+    if (! self.outputQueue) {
+        _outputDeviceID = outputDeviceID;
+        return YES;
     }
 
-    return running;
-}
-
-#pragma mark - Buffer Management
-- (void)provideAudioFrames:(OCTToxAVPCMData *)pcm
-               sampleCount:(OCTToxAVSampleCount)sampleCount
-                  channels:(OCTToxAVChannels)channels
-                sampleRate:(OCTToxAVSampleRate)sampleRate
-                fromFriend:(OCTToxFriendNumber)friendNumber
-{
-    if (self.friendNumber != friendNumber) {
-        return;
+    if ([self.outputQueue setDeviceID:outputDeviceID error:error]) {
+        _outputDeviceID = outputDeviceID;
+        return YES;
     }
-
-    int32_t len = (int32_t)(channels * sampleCount * sizeof(int16_t));
-
-    TPCircularBufferProduceBytes(&_outputBuffer, pcm, len);
-
-    if ((self.outputSampleRate != sampleRate) || (self.outputNumberOfChannels != channels)) {
-        NSError *error;
-        if (! [self updateOutputSampleRate:sampleRate channels:channels error:&error]) {
-            DDLogWarn(@"%@, error updateOutputSampleRate:%@", self, error);
-        }
+    else {
+        return NO;
     }
```

Comment:
No need for else here, you can simply `return NO;`. (it is a matter of style, but I prefer early returns without "else" part).